### PR TITLE
Refactor to decouple from relying directly on proto

### PR DESCRIPTION
**Pull request by Jefffrey**
Part of https://github.com/datafusion-contrib/datafusion-orc/issues/41

Start introducing some of our own structs/types to represent their protobuf counter part

Allows us to have more control, such as ensuring fields are actually present (since proto has them as Option usually)

### CODE REVIEW

File: src/reader/metadata.rs [hunk 1]
```diff
@@ -22,29 +22,108 @@
...
-    pub footer: Footer,
-    pub metadata: Metadata,
-    pub stripe_footers: Vec<StripeFooter>,
+    compression: Option<Compression>,
+    type_description: Arc<TypeDescription>,
+    number_of_rows: u64,
+    /// Statistics of columns across entire file
+    column_statistics: Vec<ColumnStatistics>,
+    stripes: Vec<StripeMetadata>,
+    user_custom_metadata: HashMap<String, Vec<u8>>,
```
Comment (Jefffrey) [inline src/reader/metadata.rs:55]: Doesn't contain all fields yet, but can fill in more as needed (such as rowIndexStride)

File: src/reader/metadata.rs [hunk 2]
```diff
@@ -22,29 +22,108 @@
...
+    // TODO: for now keeping this, but ideally won't want all stripe footers here
+    //       since don't want to require parsing all stripe footers in file unless actually required
+    stripe_footers: Vec<StripeFooter>,
```
Comment (Jefffrey) [inline src/reader/metadata.rs:58]: Hoping to move this out of this struct, as don't want to require parsing all stripe footers in file upfront (at least not in this metadata section)

File: src/statistics.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            Some(TypeStatistics::String {
+                min: stats.minimum().to_owned(),
+                max: stats.maximum().to_owned(),
+                sum: stats.sum(),
+            })
+        } else if let Some(stats) = &value.bucket_statistics {
+            // TODO: false count?
+            Some(TypeStatistics::Bucket {
+                true_count: stats.count[0], // TODO: safety check this
+            })
```
Comment (Jefffrey) [inline src/statistics.rs:109]: ORC spec is kinda vague on details here, been referring to https://github.com/apache/orc/blob/e1c7e4d9150820fa64b151ad942bffdf81ef34d0/c%2B%2B/src/Statistics.cc#L197

Might need further investigation

**Review by WenyXu — APPROVED** [review]
LGTM

**Pull request merged**